### PR TITLE
Add dilation option to ResNet

### PR DESCRIPTION
**Pull request by fmassa**
This PR adds support for replacing 2x2 strides in ResNet with dilation.

This functionality is useful for segmentation models, where high-resolution feature maps are desirable.

Instead of hard-coding the resolution of the feature map, we let the user specify explicitly which blocks they want the 2x2 stride to be replaced by a dilated convolution.
Given that there are only 3 blocks with 2x2 strides (if we don't consider the initial convolution, which also performs downsampling), the user needs to specify a 3-element list of booleans.

Suggestions for a better naming (instead of `replace_stride_with_dilation`) are welcome.

cc @aadcock

### CODE REVIEW

**Comment by fmassa** [comment]
I'm merging this, but let me know if anyone disagrees with something or has a better idea on how to improve this part.

**Comment by labor00** [comment]
This change is backward compatible? For now I have been using the torch.nn.apply function to directly change the parametrization of the convolution layers, replacing padding, strides and dilation?

**Comment by fmassa** [comment]
@labor00 yes, this change is backwards-compatible.

Old models and code still work the same as before, because we do not modify the strides / dilations if the parameters are the default (`None`)

**Comment by usdflt** [comment]
The repository for the gluoncv is very well documented and have lots of tests. i believe that some of the models could be ported to pytorch without hurdle.

**Comment by fmassa** [comment]
@usdflt what models do you have in mind? Note that we are adding segmentation and detection models to torchvision soon.

**Comment by usdflt** [comment]
@fmassa I was thinking of models like pspnet, yolo-v*, ssd. Is there any plans to merge the maskrccn repo within the torch vision?

**Comment by fmassa** [comment]
@usdflt the plan is to include some simplified version of faster r-cnn and mask r-cnn in torchvision
